### DISCUSSION TITLE
Allow passing "empty" values from Go SDK

### DISCUSSION THREAD

**Discussion by jedevc** [dagger/dagger#12144] [Ideas]
Suppose the following scenario:

```go
// module foo
func (f *Foo) DoFoo(
	// +optional
	// +default=true
	arg bool,
	// +optional
	arg2 string,
} error {
	// ...
}
```

```go
// module bar
func (f *Bar) DoBar() error {
	return dag.Foo().DoFoo(dagger.FooDoFooOpts{Arg: false})
}
```

As described in https://github.com/dagger/dagger/issues/7777#issuecomment-2203387161, this `false` won't apply, and we'll get the default `true` value set for `arg`.

That's because of the following logic in the generated go code: https://github.com/dagger/dagger/blob/6873b3a3960f7927c6b0971e97141ad3bc9a3e55/sdk/go/dagger.gen.go#L1092-L1095

If a value is a zero-value (e.g. `false`, `0`, `""`), then we won't include in the values sent to the server. This is done because in go **there's no way to distinguish between a value that's not provided and a value that's set to the zero value**. This issue mostly appears with default booleans of `true`, but could also appear with empty strings.

This is a pretty big issue for the Go SDK - it means that users are incentivized to just avoid options that default to `true` if they want to be called from the Go SDK (other languages don't have this issue).

We should work out what to do, since this is a pretty big stumbling block. Some options off the top of my head:

- Document this limitation somewhere - all SDKs have their quirks, so we could have a docs page where we write them all down, and we just accept that this is one of Go's.

- Explicitly handle the boolean default `true` case - in this case, invert the option for go. So in the above example, the Go SDK gets an option `NoArg`:

  ```go
  // module bar
  func (f *Bar) DoBar() error {
    	return dag.Foo().DoFoo(dagger.FooDoFooOpts{NoArg: true})
  }
  ```
  
  However, this only handles for the boolean case, we still haven't handled the similar (and more complex) `""`/`0` cases.

- Define our own "special" booleans like `dagger.False` and `dagger.True` - these would be incompatible with `bool`, but would essentially let us have a trinary value: `True`, `False` and "unset".

  However:
  1. Only solves this for booleans again.
  2. Backwards incompatible, arg types change.
  3. *Incredibly* miserable to actually use.

- Use pointers for every optional arg. This means that we can easily see the difference between a provided `&false` and an empty `nil`.

  However:
  1. You can't take the address of a literal value - so we'd need to use a pointer like `ptr(false)` (I think we already have this).
  3. This is backwards incompatible, since we change the types of args.

- Give up entirely on using structs for opts - instead allow using a `With` pattern - this is a pretty standard pattern in the go ecosystem of doing options (which we use ourselves in the engine/cli).

  ```go
  // module bar
  func (f *Bar) DoBar() error {
    	return dag.Foo().DoFoo(dagger.FooDoFooWithArg(false))
  }
  ```

  - This is very verbose. Needing to explicitly prefix `FooDoFoo` for each distinct option is quite miserable.
  - Explodes the size of go codegen - this is *probably* fine?
  - Potentially we could support this alongside our existing mechanism - currently the function signature of `DoFoo` looks like: `func (r *Foo) Export(ctx context.Context, opts ...FooDoFooOpts) (string, error)`. If we change the type of the auto-generated `Opts` struct to be a go interface, and have a struct that implements it, we could additionally have functions that implement it. This could result in calls like:
   
    ```go
    dag.Foo().DoFoo(dagger.FooDoFooOpts{Arg2: "hello"}, dagger.FooDoFooWithArg(false))
	// or even...
	dag.Foo().DoFoo(dagger.FooDoFooWithArg2("hello"), dagger.FooDoFooWithArg(false))
    ```

I think the last option here gives us some interesting opportunities - at the very least it lets us solve this problem - and potentially provides an alternative way of doing options for users who prefer the second method. I think it would be the way to go - we can do it in a fully backwards-compatible way - but it's probably going to take a bit of effort, and I'm still not a huge fan of it's huge verbosity - maybe there's something clever we could do there.

**Comment by sagikazarmark** [comment]
I think it makes sense to mention that Dagger used to have an `Optional` type that handled this situation very well. It was removed because it made the API and code more verbose.

Is this problem present with pointers as well?

```go
// module foo
func (f *Foo) DoFoo(
	// +optional
	// +default=true
	arg *bool,
} error {
	// ...
}
```

It may not be perfect, but it could be a short-term workaround.

**Comment by jedevc** [comment]
Yeah, this is kinda weird - pointers let the *callee* handle this, but there's no way for the *caller* to handle it.

I can't call `DoFoo` from another Go module to get a `false` arg (no matter how it's been declared) - at the moment at least.

> Is this problem present with pointers as well?

Woops, there is another way to handle this with pointers, edited to add this to the list of potential solutions.

**Comment by helderco** [comment]
I think this version is less verbose:
```go
dag.Foo().DoFoo(dagger.FooDoFooOpts().Arg2("hello").Arg(false))
```
Which would still allow this:
```go
dag.Foo().DoFoo(dagger.FooDoFooOpts().Arg2("hello"), dagger.FooDoFooOpts().Arg(false))
```
Probably not used inline like that, but saved into vars and composed.

**Comment by jedevc** [comment]
Ooh, I like the arg chaining a lot, definitely way less verbose, and we don't need to do the messing around with interfaces. If we want to keep backwards compat, we can't *quite* have those names, we'd have to have `WithArg` or similar.

I think the `FooDoFooOpts` structs would just look like this then:

```go
type FooDoFooOpts struct {
	Arg bool
	Arg2 string

	// we can use these set booleans to check if it was explicitly set
	// otherwise we fallback to querybuilder.IsZeroValue
	argSet bool
	arg2Set bool
}

func (opts FooDoFooOpts) WithArg(x bool) FooDoFooOpts {
	opts.Arg = x
	opts.argSet = true
}

func (opts FooDoFooOpts) WithArg2(x string) FooDoFooOpts {
	opts.Arg2 = x
	opts.arg2Set = true
}
```

Then we could have:

```go
dag.Foo().DoFoo(dagger.FooDoFooOpts{}.WithArg2("hello").WithArg(false))
// or
dag.Foo().DoFoo(dagger.FooDoFooOpts{Arg2: "hello"}.WithArg(false))
```

This would keep backwards compatibility.

**Comment by nipuna-perera** [comment]
Incredible that I didn't realize this was a limitation. Would have led to big confusion. I am closely watching this as not having default true `boolean`s would be bad for multiple reasons!

I like the chaining method if that's what you decide to go with and this will need to be well documented.

**Comment by shykes** [comment]
Can I suggest an option that may seem weird, but could have some potential? I call it "goofy/regular".

TLDR: introduce an alternative calling convention with a different DX and more features. The two conventions would co-exist, preventing any breaking changes.

- Regular convention: same as today.
- Goofy convention: encode the call in a single object, then make the call with `Callxxx`.

### Benefits

1. A clean DX for explicitly setting and getting optional arguments
2. No breaking changes. New DX is opt-in, and can be mixed and matched
3. Also solves the problem of middleware: @jedevc you could use goofy convention for GHA defaults :)

### Caller example

| Regular convention | Goofy convention  |
| -- | -- |
| `App.Test(source, dagger.AppTestOpts{All: false}))` | `App.CallTest(new(dagger.AppTestArgs).SetSource(source).SetAll(false))` |

### Callee example

The nice thing is that the callee can also choose between goofy and regular.

| Regular convention | Goofy convention  |
| -- | -- |
| `func Test(source *dagger.Directory, all bool)` | `func Test(args dagger.AppTestArgs) { source := args.GetSource(); all, ok := args.GetAll()` |

**Comment by jedevc** [comment]
We discussed in discord, but summarizing my thoughts on the goofy calling convention (for the record):

- I agree on the callee example - this would be *very* useful in lots of places, like our `gha` module. No notes, let's do it :tada:
- For the caller example:
	- Setting required args through `.Set` methods has a few down-sides: it's now possible to accidental omit a required arg (the type checker can't catch this, and instead we get a runtime error), it's entirely different from every other SDK (we've spend a lot of time with `dagger shell` to avoid this problem that we've got in `dagger call`)
	- Having separate `Call` methods and `Opts` structs generates a *lot* more code - users will see this in their IDEs, which is potentially confusing.
	- If you want to call a method with an `optional` bool, you might now need to switch calling conventions - instead of just adding one `.SetTheBoolArg(false)`.
- Adding/extending calling conventions has a maintenance cost. There are now many more code paths to follow and test: we've experienced this before when we properly supported "inline opts structs" - which is one of the primary reasons we removed that in the first place.

  That's not a reason to keep the status quo, but we should weigh up the pros/cons very carefully here - changing conventions is a lot of complexity (though I think extending is probably easier) in codes, docs, cognitive overhead, etc.

**Comment by nipuna-perera** [comment]
Any traction on this? I ran into this again today and it wasn't a pleasant experience. To find details about it I had to search Discord. Would it make sense to document here until we arrive at a conclusion? https://docs.dagger.io/api/arguments/#optional-arguments

**Comment by frizzr** [comment]
Just a simple question from a newbie here. 

So is there no _current_ workaround for having a default value of true? I'm not meaning to be a smarty-pants, but given none of this is documented (you guys **_really_** need to document this) I don't think I have any other easy way to find this out without just asking this question.

There are definitely multiple people in our company that have been directly impacted by this (just in case you need more evidence that this is an important issue.)

**Comment by nipuna-perera** [comment]
I agree. It's so easy to hit, I've done it multiple times myself without realizing.

**Comment by rrati** [comment]
I just learning dagger and I just ran into this as well.  It was not obvious what the failure was and got lucky I knew @nipuna-perera who had already run into this.  Otherwise I would be scratching my head.
